src/v/serde/protobuf/parser.cc

```diff
@@ -0,0 +1,522 @@
+                        std::make_move_iterator(elems.begin()),
+                        std::make_move_iterator(elems.end()),
+                        std::back_inserter(vec));
+                  },
+                  repeated.elements);
+            } else {
+                auto& vec = std::get<chunked_vector<T>>(repeated.elements);
+                vec.push_back(std::forward<T>(value));
+            }
+        }
+    }
+
+    void
+    stage(int32_t field_number, iobuf iobuf, const pb::Descriptor& descriptor) {
+        current_ = &state_.emplace_back(
+          field_number,
+          iobuf_parser(std::move(iobuf)),
+          std::make_unique<parsed::message>(),
+          &descriptor);
+    }
+
+    tag read_tag() {
+        auto tag_value = read_varint<uint64_t>(current_->parser);
+        constexpr uint64_t wire_mask = 0b111;
+        uint64_t wtype = tag_value & wire_mask;
+        if (wtype > static_cast<int64_t>(wire_type::max)) {
+            throw std::runtime_error(
+              fmt::format("invalid wire type: {}", wtype));
+        }
+        uint64_t field_number = tag_value >> 3ULL;
+        constexpr auto max_field_number = static_cast<uint64_t>(
+          std::numeric_limits<int32_t>::max());
+        if (field_number > max_field_number) {
+            throw std::runtime_error(
+              fmt::format("invalid field number: {}", field_number));
+        }
+        return {
+          static_cast<wire_type>(wtype), static_cast<int32_t>(field_number)};
+    }
+
+    void read_packed_elements(
+      const pb::FieldDescriptor& field, size_t amount, auto reader) {
+        if (!field.is_packable()) {
+            throw std::runtime_error(
+              fmt::format("invalid packed field: {}", field.DebugString()));
+        }
+        chunked_vector<std::invoke_result_t<decltype(reader)>> vec;
+        auto target = current_->parser.bytes_consumed() + amount;
+        while (current_->parser.bytes_consumed() < target) {
+            vec.push_back(reader());
+        }
+        append_repeated_field(field.number(), parsed::repeated(std::move(vec)));
+    }
+
+    void read_length_field(int32_t field_number) {
+        auto length = static_cast<int32_t>(
+          read_varint<uint64_t>(current_->parser));
+        if (length < 0) {
+            throw std::runtime_error(
+              fmt::format("invalid length field: (length={})", length));
+        }
+        const auto* descriptor = current_->descriptor->FindFieldByNumber(
+          field_number);
+        if (!descriptor) {
+            return;
+        }
+        switch (descriptor->type()) {
+        case pb::FieldDescriptor::TYPE_STRING:
+        case pb::FieldDescriptor::TYPE_BYTES: {
+            auto buf = current_->parser.share(length);
+            update_field(*descriptor, std::move(buf));
+            break;
+        }
+        case pb::FieldDescriptor::TYPE_MESSAGE: {
+            auto buf = current_->parser.share(length);
+            stage(
```

Comment:
nit: maybe to make expectations for callers/readers clearer, could this be called `stage_message()`?

Reply:
Done